apiserver/remoterelations/remoterelations.go

```diff
@@ -286,12 +289,106 @@ func (api *RemoteRelationsAPI) ConsumeRemoteApplicationChange(
 
 // PublishLocalRelationChange publishes local relations changes to the
 // remote side offering those relations.
-func (api *RemoteRelationsAPI) PublishLocalRelationsChange(
+func (api *RemoteRelationsAPI) PublishLocalRelationChange(
 	changes params.RemoteRelationsChanges,
 ) (params.ErrorResults, error) {
 	return params.ErrorResults{}, errors.NotImplementedf("PublishLocalRelationChange")
 }
 
+// RegisterRemoteRelations sets up the local model to participate
+// in the specified relations. This operation is idempotent.
+func (api *RemoteRelationsAPI) RegisterRemoteRelations(
+	relations params.RegisterRemoteRelations,
+) (params.ErrorResults, error) {
+	results := params.ErrorResults{
+		Results: make([]params.ErrorResult, len(relations.Relations)),
+	}
+	for i, relation := range relations.Relations {
+		if err := api.registerRemoteRelation(relation); err != nil {
+			results.Results[i].Error = common.ServerError(err)
+			continue
+		}
+	}
+	return results, nil
+}
+
+func (api *RemoteRelationsAPI) registerRemoteRelation(relation params.RegisterRemoteRelation) error {
+	// TODO(wallyworld) - do this as a transaction so the result is atomic
+	// Perform some initial validation - is the local application alive?
+
+	// TODO(wallyworld) - look up local name from offered name
+	localApplicationName := relation.OfferedApplicationName
+
+	localApp, err := api.st.Application(localApplicationName)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if localApp.Life() != state.Alive {
+		return errors.NotFoundf("application %v", localApplicationName)
+	}
+	eps, err := localApp.Endpoints()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// Does the requested local endpoint exist?
+	var localEndpoint *state.Endpoint
+	for _, ep := range eps {
+		if ep.Name == relation.LocalEndpointName {
+			localEndpoint = &ep
+			break
+		}
+	}
+	if localEndpoint == nil {
+		return errors.NotFoundf("relation endpoint %v", relation.LocalEndpointName)
+	}
+
+	// Add the remote application reference.
```

Comment:
please expand on the formulation of app name, what it refers to and why it needs to include the token